components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java

```diff
@@ -294,6 +307,70 @@ private void addEncryptedAuthSecretsToBuilder(ActionDTOBuilder actionDTOBuilder,
                         .build());
     }
 
+    private void addActionRule(ActionDTOBuilder actionDTOBuilder, String tenantDomain) throws ActionMgtException {
+
+        if (actionDTOBuilder.getActionRule() != null) {
+            try {
+                ActionMgtServiceComponentHolder.getInstance()
+                        .getRuleManagementService()
+                        .addRule(actionDTOBuilder.getActionRule().getRule(), tenantDomain);
+            } catch (RuleManagementException e) {
+                throw new ActionMgtServerException("Error while adding the Rule associated with the Action.", e);
+            }
+        }
+    }
+
+    private void loadActionRule(ActionDTOBuilder actionDTOBuilder, String tenantDomain)
+            throws ActionMgtServerException {
+
+        if (actionDTOBuilder.getActionRule() != null) {
+            try {
+                ActionRule actionRule = ActionRule.create(ActionMgtServiceComponentHolder.getInstance()
+                        .getRuleManagementService()
+                        .getRuleByRuleId(actionDTOBuilder.getActionRule().getId(), tenantDomain));
+                actionDTOBuilder.rule(actionRule);
+            } catch (RuleManagementException e) {
+                throw new ActionMgtServerException("Error while retrieving the Rule associated with the Action.", e);
+            }
+        }
+    }
+
+    private void updateActionRule(ActionDTOBuilder updatingActionDTOBuilder, ActionDTO existingActionDTO,
+                                  String tenantDomain)
+            throws ActionMgtException {
+
+        if (existingActionDTO.getActionRule() == null && updatingActionDTOBuilder.getActionRule() != null) {
+            // This means a new action rule is added when updating the action.
+            addActionRule(updatingActionDTOBuilder, tenantDomain);
+        } else if (existingActionDTO.getActionRule() != null && updatingActionDTOBuilder.getActionRule() == null) {
+            // This means the existing action rule is removed when updating the action.
+            deleteActionRule(existingActionDTO, tenantDomain);
+        } else if (existingActionDTO.getActionRule() != null && updatingActionDTOBuilder.getActionRule() !=
+                null) {  // This means the existing action rule is updated when updating the action.
+            try {
+                updatingActionDTOBuilder.getActionRule().getRule().setId(existingActionDTO.getActionRule().getId());
+                ActionMgtServiceComponentHolder.getInstance()
+                        .getRuleManagementService()
+                        .updateRule(updatingActionDTOBuilder.getActionRule().getRule(), tenantDomain);
+            } catch (RuleManagementException e) {
+                throw new ActionMgtServerException("Error while updating the Rule associated with the Action.", e);
+            }
```

Comment:
Shall we extract these to a separate method as well ?

Reply:
This method is already about rule update. so I don't think it's necessary